src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/ReadOnlyTagHelperAttributeList.cs

```diff
@@ -8,67 +8,66 @@
+            Attributes = new List<TagHelperAttribute>();
         }
 
         /// <summary>
-        /// Instantiates a new instance of <see cref="ReadOnlyTagHelperAttributeList{TAttribute}"/> with the specified
+        /// Instantiates a new instance of <see cref="ReadOnlyTagHelperAttributeList"/> with the specified
         /// <paramref name="attributes"/>.
         /// </summary>
         /// <param name="attributes">The collection to wrap.</param>
-        public ReadOnlyTagHelperAttributeList(IEnumerable<TAttribute> attributes)
+        public ReadOnlyTagHelperAttributeList(IEnumerable<TagHelperAttribute> attributes)
         {
             if (attributes == null)
             {
                 throw new ArgumentNullException(nameof(attributes));
             }
 
-            Attributes = new List<TAttribute>(attributes);
+            Attributes = new List<TagHelperAttribute>(attributes);
         }
 
         /// <summary>
-        /// The underlying collection of <typeparamref name="TAttribute"/>s.
+        /// The underlying collection of <typeparamref name="TagHelperAttribute"/>s.
         /// </summary>
-        /// <remarks>Intended for use in a non-read-only subclass. Changes to this <see cref="List{TAttribute}"/> will
-        /// affect all getters that <see cref="ReadOnlyTagHelperAttributeList{TAttribute}"/> provides.</remarks>
-        protected List<TAttribute> Attributes { get; }
+        /// <remarks>Intended for use in a non-read-only subclass. Changes to this <see cref="List{TagHelperAttribute}"/> will
+        /// affect all getters that <see cref="ReadOnlyTagHelperAttributeList"/> provides.</remarks>
+        protected List<TagHelperAttribute> Attributes { get; }
 
         /// <inheritdoc />
-        public TAttribute this[int index] => Attributes[index];
+        public TagHelperAttribute this[int index] => Attributes[index];
 
         /// <summary>
-        /// Gets the first <typeparamref name="TAttribute"/> with <see cref="IReadOnlyTagHelperAttribute.Name"/>
+        /// Gets the first <typeparamref name="TagHelperAttribute"/> with <see cref="TagHelperAttribute.Name"/>
         /// matching <paramref name="name"/>.
         /// </summary>
         /// <param name="name">
-        /// The <see cref="IReadOnlyTagHelperAttribute.Name"/> of the <typeparamref name="TAttribute"/> to get.
+        /// The <see cref="TagHelperAttribute.Name"/> of the <typeparamref name="TagHelperAttribute"/> to get.
         /// </param>
-        /// <returns>The first <typeparamref name="TAttribute"/> with <see cref="IReadOnlyTagHelperAttribute.Name"/>
+        /// <returns>The first <typeparamref name="TagHelperAttribute"/> with <see cref="TagHelperAttribute.Name"/>
         /// matching <paramref name="name"/>.
         /// </returns>
         /// <remarks><paramref name="name"/> is compared case-insensitively.</remarks>
-        public TAttribute this[string name]
+        public object this[string name]
```

Comment:
Talked with @DamianEdwards offline and the decision was to:

- Remove Setter
- Change getter to be `TagHelperAttribute`
- Add separate `Set` method(s).